### PR TITLE
Add prompt learning for T5

### PR DESCRIPTION
**Pull request by HeyyyyyyG**
# What does this PR do ?

Implement prompt learning for T5 models.

**Collection**: [NLP]

# Changelog 
- A base prompt learning dataset class (will support both T5 and GPT in the future).
- A T5-specific prompt learning dataset class.
- A base prompt learning model class (will support both T5 and GPT in the future).
- A T5-specific prompt learning model class.
- Scripts for training and inference.
- Config files for training and inference. 
- Temporarily fix a bug in T5 decode function to support prompt learning inference.

# Usage

```
python megatron_t5_prompt_learning.py --config-name=megatron_t5_prompt_learning.yaml
python megatron_t5_prompt_learning_eval.py t5_model_file=models/megatron_t5_220m_f16.nemo virtual_prompt_model_file=models/p_tuning_sent_t5.nemo
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by vadam5 — CHANGES_REQUESTED** [review]
Over all good work! I left a few comments. In addition, could you also remove all old ptuning files? This should at a minimum include these 3:

```
nlp/models/language_modeling/megatron_ptune_t5_model.py
nlp/data/glue_benchmark/t5_ptune_dataset.py
nlp/modules/common/t5_prompt_encoder.py
```

File: nemo/collections/nlp/data/language_modeling/megatron/t5_prompt_learning_dataset.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+            if len(input_ids) > self.max_seq_length:
+                input_ids = self._truncate_input(truncation_field, input_ids, taskname, doc)
+
+            # get answer ids
+            if answer_field in doc.keys():  # training and validation
+                answer_text = doc[answer_field]
+
+                # a trick to align with the data format in t5 pretraining
+                answer_text_ids = (
+                    [self.tokenizer.bos_id]
```
[the remark this answers is not in the document]
Comment (HeyyyyyyG) [reply nemo/collections/nlp/data/language_modeling/megatron/t5_prompt_learning_dataset.py:100]: I think it is necessary. The decoder input should be `<bos> A B` when the label is `A B C`.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+
+    P-tuning initializes an LSTM encoder model that generates virtual prompt
+    embeddings for every task. Each task shares the same encoder. After p-tuning
+    is compelete, the learned virtual prompts can be saved to the prompt table
+    using add_ptuned_prompts_to_prompt_table(). Thus, if a user wants to add a 
+    new virtual prompt via p-tuning, they do not need to retrain on all previous 
+    tasks. This gives p-tuning the same task flexiblity as prompt-tuning.
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
```
[the remark this answers is not in the document]
Comment (HeyyyyyyG) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:67]: I don't see how we can do that. I have put those initializations which are not dependent on the frozen model in the base model's init class. Other initializations such as the `tokenizer`, `hidden_size`, and `psuedo_token_ids` are dependent on the frozen model. They can only be initialized after we load the frozen model.

File: nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer)
+
+        self.cfg = cfg
+
+        self.load_frozen_model(cfg, trainer)
+
+        self.tokenizer = self.frozen_model.tokenizer
+
+        if self.frozen_model.cfg.precision == 16:
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:67]: You should be able to get this from `frozen_model.autocast_dtype` instead of this if-else.

File: nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+
+        # Prepare pseudo token ids for virtual/virtual prompt tokens
+        self.pseudo_tokens = get_pseudo_tokens(self.max_virtual_tokens)
+        self.tokenizer.add_special_tokens({'additional_special_tokens': self.pseudo_tokens})
+        self.pseudo_token_ids = self.tokenizer.tokens_to_ids(self.pseudo_tokens)
+        self.pseudo_token_ids_start = self.pseudo_token_ids[0]
+        self.pad_token_id = self.tokenizer.pad_id if self.tokenizer.pad_id is not None else self.tokenizer.unk_id
+        self.decoder_seq_length = cfg.get('decoder_seq_length', 40)
+
+        if self.trainer.precision == 32:
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:116]: Why are we getting both `autocast_dtype` and `float_type`? They should always be the same no?

Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py:116]: Is there a case where this can be different?

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 5]
```diff
@@ -442,7 +442,10 @@ def fwd_output_only_func(batch, model):
                 decoder_attn_mask,  # dec_attn_mask
                 None,  # token_type_ids
                 None,  # labels
-                enc_input,  # enc_hidden_states
+                None,  # enc_hidden_states
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:445]: This bit is being refactored in this PR - https://github.com/NVIDIA/NeMo/pull/4210. Do we have to add all these additional args to make it work?

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+        if inference:
+            input_embeds = self.embed_input_inference(input_ids, taskname_ids)
+        else:
+            input_embeds = self.embed_input_train(input_ids, taskname_ids)
+
+        position_embeddings = self.frozen_model.enc_dec_model.encoder_embedding.position_embeddings(position_ids)
+        encoder_input = input_embeds + position_embeddings
+
+        # Call forward on T5 model with preprocessed embeddings
+        if self.float_type == torch.float32:
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:88]: You shouldn't have to do any autocasting if precision is configured properly for the prompt learning and the frozen model. If there is a precision mismatch between pre-training and prompt-learning, you should override the precision in the frozen model before restoring. The user should only configure `trainer.precision` from which the prompt-learning model and frozen model should obtain float types.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+        # TODO: Fix this once apex patches FusedScaledMaskedSoftmax.
+        # This is a workaround for the fact that `masked_softmax_fusion` has issues with certain input sizes that may be present while finetuning.
+        t5_cfg = MegatronT5Model.restore_from(cfg.get('language_model_path'), trainer=trainer, return_config=True)
+        OmegaConf.set_struct(t5_cfg, True)
+        with open_dict(t5_cfg):
+            t5_cfg.masked_softmax_fusion = False
+            t5_cfg.megatron_amp_O2 = self.megatron_amp_o2
+            # hack to make the _GLOBAL_NUM_MICROBATCHES_CALCULATOR initialize
+            t5_cfg.micro_batch_size = cfg.get('micro_batch_size', 4)
+            t5_cfg.global_batch_size = cfg.get('global_batch_size', 4)
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:128]: This is where you want to override precision. Set `t5_cfg.precision = trainer.precision`.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+                dec_attn_mask=dec_mask,
+                token_type_ids=None,
+                labels=labels,
+                enc_hidden_states=None,
+                output_enc_hidden_only=False,
+                enc_input=encoder_input,
+            )
+        else:
+            with torch.autocast(device_type="cuda", dtype=self.float_type):
+                output = self.frozen_model.enc_dec_model(
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:102]: You shouldn't call the parent model's forward function since this will not work with pipeline parallel and will need significant refactoring to work with it. Please write another method that uses apex's `forward_backward_pipelining_without_interleaving` or `forward_backward_no_pipelining`.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+        input_embeds = self.embed_input_inference(enc_input, taskname_ids)
+
+        position_embeddings = self.frozen_model.enc_dec_model.encoder_embedding.position_embeddings(position_ids)
+        encoder_input = input_embeds + position_embeddings
+
+        predicted_token_ids, log_probs = self.frozen_model.decode(
+            tokens_enc=enc_input,
+            enc_mask=enc_mask,
+            num_tokens_to_generate=self.decoder_seq_length,
+            encoder_input=encoder_input,
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:288]: I think there is an issue on the main branch right now wherein `encoder_input` will just be ignored even if you pass it. I think @michalivne is working on a fix. Lets communicate with him about this.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+                    ]  # delete the sentinel token at the beginning of label
+                    label = self.tokenizer.ids_to_text(label)
+                    all_labels.append(label)
+
+        correct, acc = 0, None
+        if all_labels:
+            for pred, label in zip(all_preds, all_labels):
+                if pred == label:
+                    correct += 1
+            acc = correct / len(all_preds)
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:347]: If your data parallel size is > 1 at predict time, this will only compute accuracy per data parallel rank. Since you use the DistributedSampler, this will only be the accuracy on a subset of your dataset. You can either use `TorchMetrics` like I do in the t5 finetuning class or do an allgather object from all data parallel ranks like we do in NMT (see https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py#L337)

**Review by MaximumEntropy — CHANGES_REQUESTED** [review]
A bunch of comments: Maybe better communicated over a call. Lets try and chat about this next week :)

**Comment by MaximumEntropy** [comment]
Thanks for the great PR! Excited to use this for a bunch of models I'm training :)

Just wanted to add a general comment. When writing some of the initial fine-tuning class and filenames, I used "T5" everywhere and I actually meant to say "encoder-decoder". Most of what you implemented is agnostic to the pre-training approach and should work pretty much any encoder-decoder model. Can we try and rename things to make it generic to the fact that is should work with any enc-dec model (except ofcourse your sentinel addition trick).

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py (hunk 5 above)
Comment (HeyyyyyyG) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:445]: No. I just temporarily modify the args so that `enc_input` can be utilized in the `decode` function. I see the changes in that PR. If it gets merged first, I can change my codes based on it.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py (hunk 9 above)
Comment (HeyyyyyyG) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:288]: Hi @michalivne ! I slightly change the args in `get_forward_output_only_func` so that `encoder_input` can be utilized. I see you are working on a fix. Please let me know when your fix is done and merged.

File: Jenkinsfile [hunk 11]
```diff
@@ -2966,6 +2966,67 @@ pipeline {
...
+                name='/home/TestData/nlp/prompt_learning/t5_p_tuning_test' \
+                trainer.devices=1 \
+                trainer.max_steps=6 \
+                trainer.max_epochs=null \
+                model.tensor_model_parallel_size=1 \
+                model.language_model_path='/home/TestData/nlp/megatron_t5/8m/megatron_t5_8m-refactor.nemo' \  
+                model.existing_tasks=[] \
+                model.new_tasks=['rte'] \
+                model.data.train_ds=['/home/TestData/nlp/prompt_learning/rte_CI_test.jsonl'] \
+                model.data.validation_ds=['/home/TestData/nlp/prompt_learning/rte_CI_test.jsonl'] \
```
[the remark this answers is not in the document]
Comment (HeyyyyyyG) [reply Jenkinsfile:2991]: Yes. Only ~10 lines.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py (hunk 9 above)
Comment (vadam5) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:386]: @michalivne wanted to follow up on this. We are shooting to merge in Jiaqi's PR before the code freeze next week. Any update on when your PR will be merged?

**Review by vadam5 — COMMENTED** [review]
Great work! Just a few small comments, mostly on the config file.

File: examples/nlp/language_modeling/megatron_t5_prompt_learning.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        if isinstance(callback, Timer):
+            trainer.callbacks[idx] = StatelessTimer(cfg.trainer.max_time,)
+
+    # hydra interpolation does not work here as the interpolation key is lost when PTL saves hparams
+    with open_dict(cfg):
+        cfg.model.precision = cfg.trainer.precision
+
+    # load existing or init new soft prompt T5 model
+    if cfg.model.get("restore_path", None):
+        model = MegatronT5PromptLearningModel.restore_from(
```
[the remark this answers is not in the document]
Comment (HeyyyyyyG) [reply examples/nlp/language_modeling/megatron_t5_prompt_learning.py:74]: Yes

**Review by vadam5 — APPROVED** [review]
Great work Jiaqi! Let's just wait for @MaximumEntropy's approval before merging.

**Review by vadam5 — APPROVED** [review]
LGTM!

**Review by MaximumEntropy — APPROVED** [review]
LGTM! Awesome PR :)

**Review by vadam5 — APPROVED** [review]
LGTM!
